blockio-api/src/System/FS/BlockIO/API.hs

```diff
@@ -214,6 +254,23 @@ hAdviseAll hbio h advice = hAdvise hbio h 0 0 advice -- len=0 implies until the
 hDropCacheAll :: HasBlockIO m h -> Handle h -> m ()
 hDropCacheAll hbio h = hAdviseAll hbio h AdviceDontNeed
 
+{-------------------------------------------------------------------------------
+  Storage synchronisation
+-------------------------------------------------------------------------------}
+
+synchroniseFile :: MonadThrow m => HasFS m h -> HasBlockIO m h -> FsPath -> m ()
+synchroniseFile hfs hbio path = do
+    b <- FS.doesFileExist hfs path
+    if b then
+      FS.withFile hfs path (FS.ReadWriteMode FS.AllowExisting) $ hSynchronise hbio
+    else
+      throwIO $ FS.ioToFsError (FS.mkFsErrorPath hfs (FS.mkFsPath [])) ioerr
+  where
+    ioerr =
+      ioeSetErrorString
+        (mkIOError doesNotExistErrorType "synchroniseFile" Nothing Nothing)
+        ("synchroniseFile: file does not exist")
```

Comment:
We _should_ be able to get the same good quality error by modifying/wrapping the error that gets throw from `withFile` or `hSyncronise`. Hopfully that would just be a matter of catching the IOError and overriding the location field.

Reply:
Presumably the `HasBlockIO` primitives themselves should rethrow `IOError`s as `FsError`s, just like [`ioHasFS`](https://hackage.haskell.org/package/fs-api-0.3.0.1/docs/src/System.FS.IO.html#local-6989586621679092074) does. Currently, our `ioHasBlockIO` does not do that, but I've made a note for myself that I should change that

And regarding the fact that we want to rely on `withFile` to throw the error, see my previous comment https://github.com/IntersectMBO/lsm-tree/pull/482#discussion_r1867619547

Reply:
Done 8a07a26fe7eceb6f7797cee8d92ee3994918fba4